### ISSUE TITLE
InitalizeComponent / Bindings operate on DataContext from previous invocation

### ISSUE THREAD

**Issue by OliverSteinau** [dotnet/wpf#7352]
When a window is opened multiple times in a row, objects that were referenced in the `DataContext` of a **previous** invocation are altered through bindings during `InitializeComponent()`.

This occurs only when the `DataContext`  is set before the call to `InitializeComponent`.

In the attached sample project, click on `Open window`, click `OK` in the window that opens, and then on `Show values`.
Then click on `Open window` again, `OK` in the next window, then on `Show values`.

Each time the `TestWindow` is opened, a new instance of `TestClass` is created (and kept in a list); this new instance is passed via a view model into the `TestWindow`. The members `Choice1` and `Choice2` are initialized to `false` and `true` respectively, and are never modified (unless you click one of the radio buttons). But after the window was opened the _second_ time, the values in the `TestClass` object used in the _first_ invocaton are altered.

If the `DataContext` is set after the call to `InitializeComponent()`, then this bug does not occur.

Output of dotnet --info:
```
.NET SDK:
 Version:   7.0.100
 Commit:    e12b7af219

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.100\

Host:
  Version:      7.0.0
  Architecture: x64
  Commit:       d099f075e4

.NET SDKs installed:
  3.1.416 [C:\Program Files\dotnet\sdk]
  5.0.404 [C:\Program Files\dotnet\sdk]
  7.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.22 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.31 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.22 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.31 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.22 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.31 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]
```
[WpfApp2.zip](https://github.com/dotnet/wpf/files/10194237/WpfApp2.zip)

**Comment by anjalisheel-wpf** [comment]
@OliverSteinau 

Set the `DataContext` of an inner element and not on the Window / UserControl itself.

Consider the OOP design basics. Forget about WPF / XAML specifics and remember that you derive from a Window class and create a subclass of it. The contract of this class includes a public get/set property named `DataContext` which accepts any kind of object. So, if someone is replacing your `DataContext` from the outside then we will be screwed up. When you instead set the `DataContext` on the next-inner FrameworkElement inside the window, it is hosted in an environment that is owned by the window.

It shouldn't matter when you should set `DataContext` equal to view model object, before or after the call to `InitializeComponent()`, unless you rely on some bindings that are established during the call to the `InitializeComponent()`, like `ElementName` bindings.

The `InitializeComponent()` method itself locates a URI to the compiled XAML file and passes it to a `LoadComponent()` method that parses the BAML, i.e. the compiled XAML, and creates instances of the elements that you have defined in your XAML markup.

Simply setting the `DataContext` property of the window to an instance of a view model class that elements in the view bind to might as well be done after the call to the `InitializeComponent()` method. These bindings aren't resolved before the constructor returns anyway.

So usually, it does not matter but set `DataContext` after `InitializeComponents`. When `DataContextChanged` event is called, you naturally expect that components are already initialized.

Also, it's good to know whether the components can be initialized without `DataContext` and separate possible initialization issues from binding issues. If you set `DataContext` before `InitializeComponents`, the binding issues may result in an exception in `InitializeComponents`.

Make your `ViewModel` constructor very fast. Don't do any DB calls or any I/O calls, etc. You want to display the UI as soon as possible.

Make sure your `ViewModel` constructor never throws an exception. Parameter validations is OK, but just for debugging purposes. It should never happen in production.

If you need to load data into viewmodel, create separate async method called e.g. Activate(), which you will call from View's `Loaded` or `OnNavigatedTo` event.

Additionally, if you subscribe to some events in `ViewModel`, you should unsubscribe. Ideal place for subscription is Activate method, resp Deactivate to unsubscribe. If you subscribe in ViewModel's ctor, it may happen that Activate/Deactivate will never be called and you introduced memory leak.

If you feel your bindings are slowing the UI, try to use `{Binding IsAsync=True},` resp x:Bind, or try to use codebehind to set the properties in worst case.

**Comment by OliverSteinau** [comment]
You're right with all the points you mention - and still it is an error (in my opinion anyway) that the **previously** set DataContext object is modified after a new DataContext object was set. Regardless whether it was set on the window or on any component therein.
